### PR TITLE
Ry/node client

### PR DESCRIPTION
**Pull request by RaghvindYadav**
## What does this PR do?
This is the initial approach to the node client, which is helping us establish a connection with the first-spark-server.

### Related Issues & PRs
#23 

#### Key Decisions

- **Implemeneted Approach**:
  We decide to use node client for backend support.
- **Alternatives Considered**:
 We can directly integrate the backend without creating any specific package.
- **Rationale**:
  In this approach I and Holger feel it will be easy to implement for fronted developer and fast too.

## Breaking Changes

- [ ] This PR includes breaking changes
  

### Limitations & Concerns

- **Known limitations**:
- Email verification isn't working for signup because our backend currently requires either userId or authToken in the header for security reasons.
- **Edge cases**:
- **Potential risks**:

---
---

### Next Steps

- [ ] Follow-up tasks are created
  - [ ] Update user queries

### How to Test
You just need to follow signUp, signIn, signOut, forgot password and many more.

## Final Checklist
- [x] Rebased with target branch
- [x] No lint errors
- [x] QA'd in multiple browsers
- [x] Added/updated tests (if applicable)
- [x] Mobile responsive (if applicable)
- [x] Updated documentation (if applicable)
- [x] Notified the Team of this updates readiness

---

## AI Usage

- [ ] AI tools were used in development
  - **Tool used**:
    Augment code
  - **Prompts used**:
    
    - **Initial prompt**: 
- [x] No AI tools were used
  - **Reason**:
    We did manual coding because AI doesn't know more about first-spark-server. 

## References

- **Documentation**:
I wrote a document (fsdata-initialization.md) in docs directory for global node client initialisation.

### CODE REVIEW

**Comment by jonahtwalker** [comment]
Why did you delete the Linked Issues/PRs from the template? #23

Please request actual reviews from team members, this is a public repo so we are not limited to a single reviewer. This section will be removed from the template when we marge Milestone 01 into `main`. Thanks!

**Comment by jonahtwalker** [comment]
There are Next Steps listed, are you going to create an Issue for that - or is this PR going to see more work before it's ready?

If it's still a draft, please mark it as a draft. 👍🏼

File: .repomix/bundles.json
Comment (jonahtwalker) [inline .repomix/bundles.json:1]: Should this have been committed?

File: docs/fsdata-initialization.md
Comment (jonahtwalker) [inline docs/fsdata-initialization.md:1]: Thanks! 👍🏼

File: src/lib/components/email-verification.svelte [hunk 1]
```diff
@@ -14,13 +16,8 @@
...
   export let showSkipButton = false;
-
-  // Event callback props
-  export let onEmailSubmit = (data: { email: string }) => {};
-  export let onVerify = (data: { email: string; code: string }) => {};
-  export let onResend = (data: { email: string }) => {};
-  export let onBack = (data: { step: string }) => {};
-  export let onSkip = () => {};
+  export let onSkip = () => {}; // Add this prop for handling skip
+  export let onVerificationSuccess = () => {}; // Add this prop for handling successful verification
```
Comment (jonahtwalker) [inline src/lib/components/email-verification.svelte]: Do we need these generated inline comments?

File: src/lib/components/email-verification.svelte [hunk 2]
```diff
@@ -100,24 +136,38 @@
...
+      await startEmailVerification(email);
     } catch (error) {
       console.error('Error sending verification code:', error);
+      emailError = 'Failed to send verification code';
     } finally {
       loading = false;
     }
   };
 
+  const verifyEmailCode = async (code: string) => {
```
Comment (jonahtwalker) [inline src/lib/components/email-verification.svelte:140]: Should this be functional?

File: src/lib/components/user-nav.svelte [hunk 3]
```diff
@@ -3,30 +3,91 @@
...
+      currentUser = null;
+      await goto('/signin');
+    } catch (error) {
+      console.error('Error signing out:', error);
+    } finally {
+      isLoggingOut = false;
+    }
   };
+
+  function getInitials(firstName?: string, lastName?: string): string {
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:51]: This will need to be refactored to get the initials from the `username`. It is outlined in #40. Assuming the work hasn't started based on the Issue status, I will likely end up addressing this tomorrow and base it on your branch here.

File: src/lib/components/user-nav.svelte [hunk 4]
```diff
@@ -3,30 +3,91 @@
...
+    if (firstName && lastName) {
+      return (firstName[0] + lastName[0]).toUpperCase();
+    } else if (firstName) {
+      return firstName.slice(0, 2).toUpperCase();
+    } else if (lastName) {
+      return lastName.slice(0, 2).toUpperCase();
+    }
+    return 'GU';
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:59]: Why GU?

File: src/lib/components/user-nav.svelte [hunk 5]
```diff
@@ -3,30 +3,91 @@
...
+  }
 </script>
 
 <DropdownMenu.Root>
   <DropdownMenu.Trigger>
     <Button variant="ghost" class="relative h-8 w-8 rounded-full">
       <Avatar.Root class="h-9 w-9">
-        <Avatar.Image src="" alt="@shadcn" />
-        <Avatar.Fallback>SC</Avatar.Fallback>
+        {#if currentUser?.avatarUrl}
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:67]: We are not going to support uploading avatars for a while, let's remove this. They will only be initials and a color for Milestone 01.

File: src/lib/components/user-nav.svelte [hunk 6]
```diff
@@ -3,30 +3,91 @@
...
+          <Avatar.Image src={currentUser.avatarUrl} alt={`@${currentUser.userHandle}`} />
+        {:else if currentUser?.userHandle}
+          <Avatar.Fallback>
+            {getInitials(currentUser.firstName ?? undefined, currentUser.lastName ?? undefined)}
+          </Avatar.Fallback>
+        {:else}
+          <Avatar.Fallback>GU</Avatar.Fallback>
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:74]: Here again, why GU?

File: src/lib/components/user-nav.svelte [hunk 7]
```diff
@@ -3,30 +3,91 @@
...
     <DropdownMenu.Label class="font-normal">
       <div class="flex flex-col space-y-1">
-        <p class="text-sm font-medium leading-none">newstudent</p>
-        <p class="text-xs leading-none text-muted-foreground">newstudent@example.com</p>
+        {#if isLoading}
+          <p class="text-sm font-medium leading-none">Loading...</p>
+        {:else if currentUser}
+          <p class="text-sm font-medium leading-none">{currentUser.userHandle}</p>
+          <p class="text-xs leading-none text-muted-foreground">
+            {currentUser.email || 'No email'}
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:87]: A verified email is a requirement. This ProfileSnippet component is outlined in #40 as well.

File: src/lib/components/user-nav.svelte [hunk 8]
```diff
@@ -3,30 +3,91 @@
...
+          </p>
+        {:else}
+          <p class="text-sm font-medium leading-none">Guest</p>
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:90]: We don't show the AvatarMenu to unauthenticated users, why is this here?

File: src/lib/components/user-nav.svelte [hunk 9]
```diff
@@ -48,9 +109,15 @@
     <DropdownMenu.Item
       onclick={handleLogout}
       class="bg-destructive text-white focus:bg-destructive focus:text-white"
+      disabled={isLoggingOut}
     >
-      <LogOut class="mr-2 h-4 w-4" />
-      Log out
+      {#if isLoggingOut}
```
Comment (jonahtwalker) [inline src/lib/components/user-nav.svelte:116]: Please update this text to be `Logging out...`, to align with the Log Out button.

File: src/lib/components/ui/sidebar/sidebar-trigger.svelte [hunk 10]
```diff
@@ -9,7 +9,6 @@
     ref = $bindable(null),
     class: className,
     onclick,
-    ...restProps
```
Comment (jonahtwalker) [inline src/lib/components/ui/sidebar/sidebar-trigger.svelte:12]: Please just revert this change, you refactored away from the prebuilt component to something custom and lost features and functionality. All that should have changed in this file when you did make the update was an icon change to the hamburger. This is being reverted in #40

File: src/lib/services/fsdata/fsdata.ts
Comment (jonahtwalker) [inline src/lib/services/fsdata/fsdata.ts:1]: 👨🏼‍🍳 💋

File: src/lib/services/fsdata/fsdata.ts [hunk 11]
```diff
@@ -1,109 +1,29 @@
...
+import resetMyPassword from '@/services/fsdata/operations/resetMyPassword';
+import signInUser from '@/services/fsdata/operations/signInUser';
+import signInWithToken from '@/services/fsdata/operations/signInWithToken';
+import signMeOut from '@/services/fsdata/operations/signMeOut';
+import signUpUser from '@/services/fsdata/operations/signUpUser';
+import updateMyUser from '@/services/fsdata/operations/updateMyUser';
+import verifyMultiStepActionToken from '@/services/fsdata/operations/verifyMultiStepActionToken';
+import verifyMyEmail from '@/services/fsdata/operations/verifyMyEmail';
 
 const fsdata = {
```
Comment (jonahtwalker) [inline src/lib/services/fsdata/fsdata.ts]: This cannot be an `any`

File: src/lib/services/fsdata/init.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    inBrowser: true,
+    fsdata: {
+      url: import.meta.env.VITE_FSDATA_URL || 'http://localhost:8092/fsdata/api/graphql',
+      headers: {
+        [HttpHeaderName.consumer]: 'first-spark-app',
+      },
+    },
+  };
+
+  if (import.meta.env.VITE_APP_ENVIRONMENT) {
```
Comment (jonahtwalker) [inline src/lib/services/fsdata/init.ts]: Can you explain why you decided to use `import.meta.env` over `process.env`?

File: src/lib/services/fsdata/operations/findAvailableUserHandle.ts
Comment (jonahtwalker) [inline src/lib/services/fsdata/operations/findAvailableUserHandle.ts:1]: Can this please be renamed to `generateAvailableUserHandle`, or `retrieveAvailableUserHandle`? It's confusing that it's named after a handle directly but only accepts an email

File: src/lib/services/fsdata/operations/updateMyUser.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import clientStore from '@/services/fsdata/clientStore';
+import { type MyUser } from '@baragaun/bg-node-client';
+
+const updateMyUser = async (changes: Partial<MyUser>): Promise<MyUser | null> => {
+  const client = clientStore.getClient();
+  if (!client) {
+    console.error('fsdata.findMyUser failed: client not initialized.');
```
Comment (jonahtwalker) [inline src/lib/services/fsdata/operations/updateMyUser.ts]: Is this error correct?

File: src/lib/services/fsdata/operations/updateMyUser.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    // todo: show error to user
+    return null;
+  }
+
+  const result = await client.operations.myUser.updateMyUser(changes);
+
+  if (result.error || !result.object?.id) {
+    console.error('createMultiStepAction failed.', result.error);
```
Comment (jonahtwalker) [inline src/lib/services/fsdata/operations/updateMyUser.ts]: Does this do anything more than updating the User object?

File: src/lib/services/fsdata/operations/signUpUser.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    console.error('fsdata.signUpUser failed: client not initialized.');
+    // todo: show error to user
+    return null;
+  }
+
+  const result = await client.operations.myUser.signUpUser({
+    userHandle,
+    email,
+    password,
+    isTestUser: true,
```
Comment (jonahtwalker) [inline src/lib/services/fsdata/operations/signUpUser.ts]: Why do we need this? Should they be `env` variables?

File: src/routes/reset-password/+page.svelte [hunk 16]
```diff
@@ -9,15 +9,24 @@
...
+  import { onDestroy } from 'svelte';
+  import { goto } from '$app/navigation';
+  import { authStore } from '$lib/components/nav-bar.svelte';
 
   let identifier = ''; // for email or username
   let loading = false;
   let error = '';
-  let emailSent = false;
+  let verificationCode = '';
+  let step = 'email'; // 'email' or 'verify'
```
Comment (jonahtwalker) [inline src/routes/reset-password/+page.svelte:21]: Could this be actually restricted to either?

File: src/routes/signin/+page.svelte [hunk 17]
```diff
@@ -56,11 +58,17 @@
...
     try {
       if (loginMethod === 'password') {
-        // TODO: Implement your password signin logic here
-        await new Promise((resolve) => setTimeout(resolve, 1000)); // Simulate API call
-        localStorage.setItem('authToken', 'your-auth-token');
-        authStore.set({ isAuthenticated: true }); // Update auth store
-        await goto('/');
+        const user = await fsdata.signInUser(
+          identifier,
+          identifier.includes('@') ? UserIdentType.email : UserIdentType.userHandle,
```
Comment (jonahtwalker) [inline src/routes/signin/+page.svelte]: I don't think this can be that simple. It is common practice (that we'll likely reinforce as well) to refer to a username or handle with an @, just as it is here in Git - `@RaghvindYadav`. I think it's just a likely point of friction with a combined input. Maybe I'm crazy?

I think it'd be better if we check for:
1. contains a `@`
2. contains chars before the `@`
3. contains a `.` after the `@`

Else, it's a username or handle.

@shadowbrush ?

File: src/routes/signin/+page.svelte [hunk 18]
```diff
@@ -56,11 +58,17 @@
...
+          password,
+        );
+        if (user) {
+          authStore.set({ isAuthenticated: true });
+          await goto('/');
+        } else {
+          error = 'Invalid credentials. Please try again.';
+        }
       } else {
         await handleMagicLinkSignIn();
```
Comment (jonahtwalker) [inline src/routes/signin/+page.svelte:71]: Since we have verification codes, we can implement the token log in method, right?

File: src/routes/signup/+page.svelte [hunk 19]
```diff
@@ -76,10 +39,23 @@
...
-      await new Promise((resolve) => setTimeout(resolve, 1000)); // Simulate API call
-      localStorage.setItem('authToken', 'your-auth-token');
-      authStore.set({ isAuthenticated: true }); // Update auth store
+      // First check if username is available
+      const isUsernameAvailable = await fsdata.isUserIdentAvailable(
+        username,
+        UserIdentType.userHandle,
+      );
+      if (!isUsernameAvailable) {
+        usernameError = 'This username is already taken';
```
Comment (jonahtwalker) [inline src/routes/signup/+page.svelte]: Please replace this with `This username is unavailable.` Until we have some moderation sorted out, I think we should keep it as ambiguous as possible as to why it's disallowed while we handle profanity and the like without an explicit policy.

File: src/routes/reset-password/+page.svelte
Comment (jonahtwalker) [inline src/routes/reset-password/+page.svelte:1]: This needs to be refactored. Somehow it's reverted back to the old validation rules from Holger's changes.

File: src/routes/signup/+page.svelte [hunk 20]
```diff
@@ -76,10 +39,23 @@
   const handleSignupSubmit = async () => {
     loading = true;
     try {
-      // TODO: Implement your signup logic here
-      await new Promise((resolve) => setTimeout(resolve, 1000)); // Simulate API call
-      localStorage.setItem('authToken', 'your-auth-token');
-      authStore.set({ isAuthenticated: true }); // Update auth store
+      // First check if username is available
+      const isUsernameAvailable = await fsdata.isUserIdentAvailable(
```
Comment (jonahtwalker) [inline src/routes/signup/+page.svelte]: This should be the second time that we confirm it's availability. The User expects to know if it's available prior to submitting the form and facing an error. Further, we should make sure it is still available before we do create the User on submission, so as to not show the user an error after we said it was green-lit.

File: src/routes/signup/+page.svelte [hunk 21]
```diff
@@ -220,22 +194,70 @@
                   >
                     At least 8 characters
                   </li>
+                  <li
+                    class:text-destructive={!/[A-Z]/.test(password)}
```
Comment (jonahtwalker) [inline src/routes/signup/+page.svelte]: This needs to be refactored. Somehow it's reverted back to the old validation rules from Holger's changes.

**Review by jonahtwalker — CHANGES_REQUESTED** [review]
It seems like there are missing files, like `package.json`. Is this actually ready for review?  

Does something need to be clarified before we can implement the errors for the user when our client doesn't connect for an operation? What about the listener errors? This PR should definitely come with setup tests.

File: src/lib/components/email-verification.svelte (hunk 1 above)
Comment (shadowbrush) [reply src/lib/components/email-verification.svelte]: I'd say less is more. The names of the handlers are descriptive enough, so I'd delete the comments.

**Comment by RaghvindYadav** [comment]
> There are Next Steps listed, are you going to create an Issue for that - or is this PR going to see more work before it's ready?

What you suggest? Should I use new branch to overcome this conflict?

File: .repomix/bundles.json
Comment (RaghvindYadav) [reply .repomix/bundles.json:1]: I tried to remove it permanently, but it every times left. I will remove it.

File: src/lib/components/email-verification.svelte (hunk 2 above)
Comment (RaghvindYadav) [reply src/lib/components/email-verification.svelte:140]: Holger is working on the backend to enable email verification for non-existing users. I will modify it later. I also discussed about it in description.

File: src/routes/signin/+page.svelte (hunk 18 above)
Comment (RaghvindYadav) [reply src/routes/signin/+page.svelte:71]: Actually not, Now we can use any six digit number to bypass the verification. I will create a new issue for this.

File: src/routes/reset-password/+page.svelte
Comment (RaghvindYadav) [reply src/routes/reset-password/+page.svelte:1]: You're correct, and I also brought it up in the standup that the current reset flow is the older version. I intentionally pushed it so everyone could see the results with the current backend. I'm working on learning backend development so I can assist Baragaun.

File: src/routes/signup/+page.svelte (hunk 20 above)
Comment (RaghvindYadav) [reply src/routes/signup/+page.svelte]: Ah, good catch. I will update the code.

File: src/routes/signup/+page.svelte [hunk 22]
```diff
@@ -76,10 +39,23 @@
...
+        username,
+        UserIdentType.userHandle,
+      );
+      if (!isUsernameAvailable) {
+        usernameError = 'This username is already taken';
+        loading = false;
+        return;
+      }
+
+      // If username is available, proceed with signup
```
Comment (shadowbrush) [inline src/routes/signup/+page.svelte]: The app should call `signUpUser` right after the email has been entered and the user clicks on "Continue". That is when we create the user account.

File: src/routes/signup/+page.svelte (hunk 19 above)
Comment (shadowbrush) [reply src/routes/signup/+page.svelte]: I made this change in the code, after you got a thumbsup from Ragvind without any further followup. @RaghvindYadav This is not OK. Don't give a thumbsup without doing what the requester asks. If we see a thumbsup we assume you have taken care of the request.

File: src/routes/signup/+page.svelte (hunk 20 above)
Comment (shadowbrush) [reply src/routes/signup/+page.svelte]: Ideally, we keep track of which values have already been confirmed as available. We did that in the Mimble app.

File: src/routes/signin/+page.svelte (hunk 18 above)
Comment (shadowbrush) [reply src/routes/signin/+page.svelte:71]: Why can't we implement this? The backend fully supports `signInWithToken`. Just like Reddit, we will send out an email with the token, and the user has to enter the token into the app. `bg-node-client` already fully supports it.

File: src/routes/signin/+page.svelte (hunk 17 above)
Comment (shadowbrush) [reply src/routes/signin/+page.svelte]: I would suggest that we don't allow usernames that contain an `@`. We can always display the username with an `@` in front, if that's what we want to do. Let's discuss it. In the Mimble app, we did that.

**Review by shadowbrush — CHANGES_REQUESTED** [review]

File: src/lib/components/email-verification.svelte (hunk 2 above)
Comment (shadowbrush) [reply src/lib/components/email-verification.svelte:140]: This is fully available.

**Comment by shadowbrush** [comment]
The app in this PR has 3(!) validatePassword functions in:

* `reset-password/+page.svelte`
* `account-settings.svelte`
* `signup/+page.svelte`

I know, you are using AI to generate code. But something like the password validation needs implemented once and only once. How did that happen? Did you think it's OK to have 3 implementation of the same thing? 

I consolidated this to `src/helpers/passwordHelpers.ts`. Not sure this location is best-practice for Svelte, so feel free to move it around.

**Comment by shadowbrush** [comment]
I rebased the branch, so be careful pulling it, if you have local changes.

**Comment by shadowbrush** [comment]
I upgraded the data client `"@baragaun/bg-node-client": "^0.0.8"`. The issue with Winston should be resolved.

File: src/routes/signup/+page.svelte (hunk 20 above)
Comment (jonahtwalker) [reply src/routes/signup/+page.svelte]: Ah! That's a good idea! Like a temporary reservation?

**Comment by jonahtwalker** [comment]
> > There are Next Steps listed, are you going to create an Issue for that - or is this PR going to see more work before it's ready?
> 
> What you suggest? Should I use new branch to overcome this conflict?

I'm not aware of a conflict to overcome, here - sorry. I was just asking if you are using the Next Steps in the PR description as a to-do list, or as a list of work that needs to be taken care of once this is merged.

I'll revisit the template again this week to try and simplify it. 👍🏼

**Comment by jonahtwalker** [comment]
> The app in this PR has 3(!) validatePassword functions in:
> 
> * `reset-password/+page.svelte`
> * `account-settings.svelte`
> * `signup/+page.svelte`
> 
> I know, you are using AI to generate code. But something like the password validation needs implemented once and only once. How did that happen? Did you think it's OK to have 3 implementation of the same thing?
> 
> I consolidated this to `src/helpers/passwordHelpers.ts`. Not sure this location is best-practice for Svelte, so feel free to move it around.

@RaghvindYadav In my PR I have this under `lib/utils/validation.ts`, if you do sort out what best practice is one way or another, will you let us know? I was thinking we could probably put it right alongside the updatepasswordform component when that gets done, as well.

File: src/routes/signin/+page.svelte (hunk 17 above)
Comment (jonahtwalker) [reply src/routes/signin/+page.svelte]: My mistake, I thought I was clear here. I was referring to the email validation only, I totally agree that usernames should **not** contain @

File: src/routes/signup/+page.svelte (hunk 21 above)
Comment (shadowbrush) [reply src/routes/signup/+page.svelte]: I refactored it. There is now a `src/helpers/passwordHelpers.ts`. Not sure that this is best-practice, but at least it's consolidated into one file now.

File: src/routes/signup/+page.svelte (hunk 20 above)
Comment (shadowbrush) [reply src/routes/signup/+page.svelte]: > Ah! That's a good idea! Like a temporary reservation?

Not really. Just let's say you entered `big-fish` and it's not available, the app shows an error. When the user tries other usernames and then tries `big-fish` again, we don't even have to ask the backend.

**Comment by shadowbrush** [comment]
I made some more changes here:

* Moved the data stuff to `src/lib/services/dataProvider`
* Renamed all `fsdata` to `dataProvider`
* Split the discussion around the initialization of `BgNodeClient` into `docs/tech-decisions/data-provider-initialization.md`, but left the first document in place. It needs to explain how the app can access the data provider.

**Comment by shadowbrush** [comment]
We probably need to remove all use of `authStore` and start using `dataProvider.isSignedIn`.

Also, the app should not save or access the `authToken` at all. That's now entirely in `BgNodeClient`.

File: src/routes/signup/+page.svelte (hunk 19 above)
Comment (RaghvindYadav) [reply src/routes/signup/+page.svelte]: okay noted

File: src/routes/signup/+page.svelte [hunk 23]
```diff
@@ -3,32 +3,17 @@
   import { Input } from '$lib/components/ui/input';
```
Comment (shadowbrush) [inline src/routes/signup/+page.svelte:3]: This needs to be split into the pages that are shown in the sequence of the onboarding. Each page will be its own component, and probably have more child components. Don't build a component that shows all pages, depending on the progress during onboarding!

**Comment by shadowbrush** [comment]
Some general comments:

* Try to name functions and variables as close to the thing that they represent. And try to keep it close to the model that they are handling. We have a `User` model and we have a `MyUser` model. When you create a variable that represents a `MyUser` object, name it `myUser`, not `user`. When you create a wrapper around `signUpUser`, give it the same name, not `signUp`
* I removed `src/lib/services/dataProvider`, and only moved the listeners into `src/lib/context/`. I don't know if they are going to be useful. If not, just delete them.
* Type all function returns: `myFunction(): ThisType`

**Comment by shadowbrush** [comment]
Also, I saw you were loading the `MyUser` in the constructor of the context. In general, avoid side effects in constructors, especially if you don't have much control over when the class is instantiated. This is especially true for things like context that should load fast and not increase the loading time of the app.

**Comment by jonahtwalker** [comment]
@RaghvindYadav Can this PR be closed?

**Comment by RaghvindYadav** [comment]
> @RaghvindYadav Can this PR be closed?

Just give us this week, I am hoping data client stuff is going to finish.

**Comment by RaghvindYadav** [comment]
I am closing this PR because these changes are included in #44 with some refractors.
